pallets/asset-registry/src/migration.rs

```diff
@@ -7,81 +7,170 @@
 
 	pub fn pre_migrate<T: Config>() {
-		assert_eq!(StorageVersion::get::<Pallet<T>>(), 0, "Storage version too high.");
+		assert_eq!(StorageVersion::get::<Pallet<T>>(), 1, "Storage version too high.");
 
 		log::info!(
 			target: "runtime::asset-registry",
 			"Asset Registry migration: PRE checks successful!"
 		);
 	}
 
-	pub fn migrate<T: Config>() -> Weight {
+	pub fn migrate<T: Config<AssetId = u32>>() -> Weight {
+		log::info!(
+			target: "runtime::asset-registry",
+			"Running migration to v2 for Asset Registry"
+		);
+
 		log::info!(
 			target: "runtime::asset-registry",
-			"Running migration to v1 for Asset Registry"
+			"Migrating Assets storage"
 		);
 
 		let mut i = 0;
-		Assets::<T>::translate(
-			|_key,
-			 OldAssetDetails {
-			     name,
-			     asset_type,
-			     existential_deposit,
-			     locked: _,
-			 }| {
-				i += 1;
-				Some(AssetDetails {
-					name,
-					asset_type,
-					existential_deposit,
-					xcm_rate_limit: None,
-				})
-			},
+		let mut v2_assets_details = Vec::<(
+			<T as crate::Config>::AssetId,
+			AssetDetails<<T as crate::Config>::AssetId, <T as crate::Config>::StringLimit>,
+		)>::new();
+		for (k, v) in v1::Assets::<T>::iter() {
+			i += 1;
+			let (symbol, decimals) = if let Some(meta) = v1::AssetMetadataMap::<T>::get(k) {
+				(Some(meta.symbol), Some(meta.decimals))
+			} else {
+				(None, None)
+			};
+
+			v2_assets_details.push((
+				k,
+				AssetDetails {
+					name: Some(v.name),
+					asset_type: v.asset_type,
+					existential_deposit: v.existential_deposit,
+					symbol,
+					decimals,
+					xcm_rate_limit: v.xcm_rate_limit,
+					//All assets created before this are sufficient
+					is_sufficient: true,
+				},
+			));
+		}
+
+		for (k, v) in v2_assets_details {
+			i += 1;
+			Assets::<T>::insert(k, v);
+			log::info!(
+				target: "runtime::asset-registry",
+				"Migrated asset: {:?}", k
+			);
+		}
+
+		//This assumes every asset has metadata and each metadata is touched.
+		i += i;
+		let _ = v1::AssetMetadataMap::<T>::clear(u32::MAX, None);
+
+		log::info!(
+			target: "runtime::asset-registry",
+			"Migrating AssetLocations storage"
 		);
 
-		StorageVersion::new(1).put::<Pallet<T>>();
+		for k in v1::AssetLocations::<T>::iter_keys() {
+			i += 1;
+
+			AssetLocations::<T>::migrate_key::<Twox64Concat, <T as crate::Config>::AssetId>(k);
 
+			log::info!(
+				target: "runtime::asset-registry",
+				"Migrated asset's location: {:?}", k
+			);
+		}
+
+		StorageVersion::new(2).put::<Pallet<T>>();
 		T::DbWeight::get().reads_writes(i, i)
 	}
 
 	pub fn post_migrate<T: Config>() {
-		assert_eq!(StorageVersion::get::<Pallet<T>>(), 1, "Unexpected storage version.");
+		for a in Assets::<T>::iter_keys() {
+			let _ = Assets::<T>::get(a).expect("Assets data must be valid");
+		}
+
+		for l in AssetLocations::<T>::iter_keys() {
+			let _ = AssetLocations::<T>::get(l).expect("AssetLocations data must be valid");
+		}
+
+		assert_eq!(v1::AssetMetadataMap::<T>::iter().count(), 0);
```

Comment:
```suggestion
		assert_eq!(v1::AssetMetadataMap::<T>::iter().count(), 0);
		assert_eq!(v1::Assets::<T>::iter().count(), 0);
```

Reply:
`v1::Assets` and latest `Assets` is the same storage so `v1::Assets` won't be empty

Reply:
the hasher changes

Reply:
aah, the iterator cannot distinguish between different hashers, ok